### PR TITLE
Extension improvements (kube version check + annotations persistence)

### LINKED ISSUE #9132: Warn users when upgrading a kubernetes version will break an installed extension
1) When an extension does not run on a given Kubernetes version the UI should warn the user when upgrading to that Kubernetes version.

2) this is also to look into extensions installed, so that when we update the k8s version, we shouldn't load the extension(s) if they are incompatible with the current k8s version

**Update 28-09-2023**

_Blocked by backend:_
https://github.com/rancher/rancher/issues/42982

_Backend enhancement needed for `ui-plugin-operator`:_
https://github.com/rancher/rancher/issues/42983

### PR DESCRIPTION
**Pull request by aalves08**
Fixes #9132 

- prevent extension load if `local` cluster kube version is not compatible with `uiPlugin` kube version
- improve kube version incompatibility check mechanism for extensions (it was flawed before because it assumed we had the versions available to cross-check it, but if the kube version is incompatible then the extension operator doesn't even load it from the backend, making it non-existent for the cross-check)
- add banner to extensions screen when there's a kube version incompatibility with any of the installed extensions
- add extension chart version annotations to the `uiPlugin` resource on extension installation

**To test**
 - install `elemental` extension (latest version)
 - check that the `uiPlugin` resource created for `elemental` has the annotations from the chart version under `spec.plugin.metadata`
 - quick test for new kube version incompatibility banner: edit https://github.com/rancher/dashboard/blob/master/shell/pages/c/_cluster/uiplugins/index.vue#L127 after install to `v1.29.2` so that the current installed version of `elemental` becomes "incompatible" with the current kube version
 - quick test for not loading extension: edit  `currKubeVersion` variable in `plugin.js` file to `v1.29.12+k3s1`so that the `local` cluster kube version is greater than `v1.28` (`elemental` default)

**Screenshots**
Screenshot 2023-09-26 at 18 24 31

Screenshot 2023-09-01 at 13 58 01

### CODE REVIEW

**Review by codyrancher — DISMISSED** [review]

**Comment by aalves08** [comment]
@richard-cox , I've fixed the e2e issue. Do you think we can still move forward with this feature for `v2.8.0`?

**Comment by richard-cox** [comment]
This is still a 2.8.0 thing. I'd like to give a quick review

**Review by richard-cox — CHANGES_REQUESTED** [review]
I didn't get to test functional stuff because we might have some issues, specifically fetching the local cluster to get it's kube version
- user may not have permissions to see the local cluster
- hopefully we should be able to load plugins on the log in screen in 2.8.0. at this point fetching the local cluster is impossible

Couple of options
- we ask that the local kube version be added to the `rancherversion` request (possible security issue?)
- we ask the response to `http:ui-plugin-operator:80/proxy/index.json` to contain info on whether they're valid or not

File: shell/assets/translations/en-us.yaml
Comment (aalves08) [reply shell/assets/translations/en-us.yaml:4172]: updated

File: shell/pages/c/_cluster/uiplugins/index.vue
Comment (aalves08) [reply shell/pages/c/_cluster/uiplugins/index.vue:459]: updated

File: shell/plugins/plugin.js
Comment (aalves08) [reply shell/plugins/plugin.js:35]: Cool. Thanks. It's only the `local` cluster needed. Will adjust accordingly

Comment (aalves08) [reply shell/plugins/plugin.js:35]: True... I'll have to examine the whole logic to check what is the contingency plan for this 🤔 Will present my findings or a solution

File: shell/pages/c/_cluster/uiplugins/index.vue
Comment (aalves08) [reply shell/pages/c/_cluster/uiplugins/index.vue:459]: > * `neu?.length === 1` - this only works if there's a single, non-built in plugin. if there's more than one it will never fire?
> * what happens with `latestInstalledPlugin` if two are installed at the same time?
> * from what i can tell it's persisting (plugin.save) package annotations in the ui-plugin metadata. this doesn't feel right. shouldn't the plugin's package information be available at other points?

Replying to your comments:
1) you are right. I don't know how I missed this. Will adjust accordingly.

2) First point: we cannot install two at the same time. The logic that copies the annotations are tied to a watcher, which will fire every time a plugin is installed, so it only runs once per install. Once it runs on this watcher, it's reset to `undefined`. It's again populated once an install occurs again. So, it should all be covered and work as expected.

3) It's the only way we can persist the data as it's the only resource available to us that is tied to plugins. Afaik, there's no other alternative and it's only metadata outside of the scope of the actual `uiplugins.metadata` field. This is something that is going to be useful to us to do more checks in the future.

File: shell/plugins/plugin.js
Comment (aalves08) [reply shell/plugins/plugin.js:65]: Will adjust code to contemplate that scenario. Thanks for the feedback.

File: shell/plugins/plugin.js
Comment (aalves08) [reply shell/plugins/plugin.js:35]: Well, I've looked into this and a user needs to be a `local` `Cluster Owner` in order to manage extensions. In that case, we got permission to get the kube version.
Worst case scenario is a standard user that doesn't have access to local cluster will **still** load extensions like he did before and they won't benefit from this update. Not ideal but no harm here (for now... we would need to create a follow-up issue).

Logic implemented covers the scenario where we it doesn't have access to local cluster.

If you are okay with this @richard-cox , we can resolve this conversation. Will leave it up to you.

**Comment by aalves08** [comment]
> I didn't get to test functional stuff because we might have some issues, specifically fetching the local cluster to get it's kube version
> 
> * user may not have permissions to see the local cluster
> * hopefully we should be able to load plugins on the log in screen in 2.8.0. at this point fetching the local cluster is impossible
> 
> Couple of options
> 
> * we ask that the local kube version be added to the `rancherversion` request (possible security issue?)
> * we ask the response to `http:ui-plugin-operator:80/proxy/index.json` to contain info on whether they're valid or not

Well, that's a good point @richard-cox . Ideal scenario, imho, would be to add it to the `rancherversion` request. It would make life so much easier.
The second option is not ideal, imo, as if we updated the local cluster kube version, we would need to ensure we updated all of the UIPlugin resources with that info. But it's definitely doable and probably quicker than waiting for something to change on the backend side. 

This can tracked in a separate issue, imo. What do you think @richard-cox ?

File: shell/plugins/plugin.js
Comment (aalves08) [reply shell/plugins/plugin.js:65]: I believe these were never captured as the `try-catch` needed to be on the `allSettled` part (here we are just queueing up promises).

https://github.com/rancher/dashboard/blob/master/shell/plugins/plugin.js#L64

File: shell/assets/translations/en-us.yaml
Comment (richard-cox) [inline shell/assets/translations/en-us.yaml:4177]: there was a typo in my suggestion `property` --> `properly`

File: shell/pages/c/_cluster/uiplugins/index.vue
Comment (richard-cox) [reply shell/pages/c/_cluster/uiplugins/index.vue]: we need to run through this together. also spotted if the user navigates away from the page the extra information is never applied

File: shell/assets/translations/en-us.yaml
Comment (richard-cox) [reply shell/assets/translations/en-us.yaml]: there was a typo in my suggestion `property` --> `properly`

File: shell/plugins/plugin.js
Comment (richard-cox) [reply shell/plugins/plugin.js]: I'm not sure this is the best path forward. Admins will then not see extensions that their users can

File: shell/plugins/plugin.js
Comment (richard-cox) [reply shell/plugins/plugin.js]: the advice to check model kubernetesVersion depends on the type of resource fetched. there'll be only one type, so no need for checking for different ones.

if the resource needs changing in to a model we should be able to classify it to access the kubernetesVersion properties

File: shell/plugins/plugin.js
Comment (richard-cox) [reply shell/plugins/plugin.js:86]: there was no allSettled before, so the exception / failure would be caught by the try block

```
  try {
      const res = await context.store.dispatch('management/request', {
        url:                  `${ UI_PLUGIN_BASE_URL }/index.json`,
        method:               'GET',
        headers:              { accept: 'application/json' },
        redirectUnauthorized: false,
      });
```

File: shell/assets/translations/en-us.yaml
Comment (aalves08) [reply shell/assets/translations/en-us.yaml]: fixed

File: shell/plugins/plugin.js
Comment (aalves08) [reply shell/plugins/plugin.js]: classify `store.getters['management/classify'](obj)` and check for `kubernetesVersion` model property

**Comment by aalves08** [comment]
will be closed as we are dependent on backend and work done on this PR would fundamentally change. Even if it's closed, we still have this as record of the work done, which is a good basis for a new PR when backend get's the time to unblock UI)

**Pull request closed without merging**
